File: BasicCalculator.py
'''
TC: O(n)
SC: O(1)
'''
class Solution:
    def calculate(self, s: str) -> int:
        stack = []
        curr = 0
        lastSign = '+'
        for i, character in enumerate(s):
            if character == ' ':
                continue
            if character.isdigit():
                curr = curr*10 + int(character)
            else:
                if character == '+' or character == '-':
                    curr = curr if lastSign == '+' else -curr
                    stack.append(curr)
                    curr = 0
                    lastSign = character
                elif character == '(':
                    stack.append(lastSign + character)
                    lastSign = '+'
                elif character == ')':
                    if lastSign == '-':
                        curr = -curr
                        lastSign = '+'
                    while stack[-1] != '+(' and stack[-1] != '-(':
                        curr = curr + stack.pop()
                    if stack[-1] == '-(':
                        lastSign = '-'
                    stack.pop()
        curr = curr if lastSign == '+' else -curr
        while stack:
            curr += stack.pop()
        return curr
s = Solution()
print(s.calculate("1 + 1"))
print(s.calculate(" 2-1 + 2 "))
print(s.calculate("(1+(4+5+2)-3)+(6+8)"))
print(s.calculate("- (3 + (4 + 5))"))

File: Basic Calculator.java
//  Time Complexity: O(n)
//  Space Complexity: O(n)
class Solution {
    public int calculate(String s) {
        int result = 0;
        int curr = 0;
        int lastSign = 1;
        int open_brace = Integer.MAX_VALUE;
        Stack<Integer> stack = new Stack<>();

        for(int i=0; i<s.length(); i++){
            char c = s.charAt(i);
            if(Character.isDigit(c)){
                curr = curr * 10 + c - '0';
            }else if(c == '('){
                stack.push(lastSign);
                stack.push(open_brace);
                curr = 0;
                lastSign = 1;
            }else if(c == ')'){
                stack.push(curr * lastSign);
                int temp = 0;
                while(stack.peek() != open_brace){
                    temp += stack.pop();
                }
                stack.pop();
                stack.push(temp * stack.pop());
                curr = 0;
                lastSign = 1;
            }else if(c =='+'){
                stack.push(curr * lastSign);
                curr = 0;
                lastSign = 1;
            }else if(c =='-'){
                stack.push(curr * lastSign);
                curr = 0;
                lastSign = -1;
            }
        }

        if(curr != 0) stack.push(curr * lastSign);

        while(!stack.isEmpty()){
            result += stack.pop();
        }

        return result;

    }
}